firmware/main.cpp

```diff
@@ -0,0 +1,140 @@
+// This can likely be fixed if we need more than 2 independent servos.
+PwmOut servos[NUM_SERVOS] = {PwmOut(PA_0), PwmOut(PA_1)};
+
+// Servo positions when DRS is not active (positions in uS, most servos expect 1000-2000 uS range)
+int servo_pos_no_drs[NUM_SERVOS] = {1000, 1000};
+// Servo positions when DRS is active
+int servo_pos_drs[NUM_SERVOS] = {2000, 2000};
+
+
+// Event queue for handling I/O triggered by an ISR (e.g. printing, sending CAN messages)
+EventQueue queue(32 * EVENTS_EVENT_SIZE);
+Thread t;
+
+// Send out a heartbeat message over CAN
+void can_hb() {
+    if(can.write(CANMessage(BRAKE_LIGHT, &counter, 1))) {
+        counter++;
+    } else {
+        printf("Failed to send CAN Heartbeat message :(\n");
+    }
+}
+
+// ISR function - schedules the can heartbeat function to run outside the ISR
+ void can_hb_handler() {
+    queue.call(can_hb);
+}
+
+// Handle CAN frames from the steering wheel
+void steering_wheel_received(CANMessage& msg) {
+    // Read the value of the DRS button from the steering wheel
+    unsigned char drs = msg.data[STEERING_WHEEL_DRS_BYTE];
+
+    printf("Steering wheel DRS setting: %02X\n", (unsigned int)drs);
+
+    // Write to the servos
+    if(drs){
+        for(int i = 0; i < NUM_SERVOS; i++){
+            servos[i].pulsewidth_us(servo_pos_drs[i]);
+        }
+    } else {
+        for(int i = 0; i < NUM_SERVOS; i++){
+            servos[i].pulsewidth_us(servo_pos_no_drs[i]);
+        }
+    }
+}
+
+// Handle CAN frames from the brake pressure sensor
+void brake_pressure_received(CANMessage& msg) {
+    unsigned char pressure = msg.data[BRAKE_PRESSURE_BYTE];
+    if(pressure > BRAKE_LIGHT_PRESSURE_CUTOFF) {
+        led.write(brightness);
+    } else {
+        led.write(0.0f);
+    }
+}
+
+void cycle_brightness() {
+    // Incrementally decrease brightness until it's below the min, then cycle back to max.
+    brightness *= 0.5f;
+    if(brightness < MIN_BRIGHTNESS) {
+        brightness = 1.0f;
+    }
+
+    // If the brake light is currently on, update its brightness.
+    if(led.read() != 0) {
+        led.write(brightness);
+    }
+}
+
+// Handle received can frames
+void can_received() {
+    CANMessage received;
+    // Read newly received CAN messages until there aren't any
+    while(can.read(received)){
+        if(received.id == STEERING_WHEEL_ID) {
+            steering_wheel_received(received);
+        } else if(received.id == BRAKE_PRESSURE_ID) {
+            brake_pressure_received(received);
+        } else if(received.id == BRAKE_LIGHT) {
+            //  Received message on brake light ID to change brightness
+            cycle_brightness();
+        }
+    }
+}
+
+// ISR function which schedules handling of received CAN frames
+void can_received_handler() {
+    queue.call(can_received);
+}
+
+int main()
+{
+    printf("Entered main\n");
+
+    // Initialize servos
+    for(auto & servo : servos) {
```

Comment:
For development and testing purposes, it may also be worthwhile to add a header to disable servos, as that’s secondary functionality. You may want to add a bit of error handling here, it looks like if this fails there may be a crash.